### PR TITLE
Added OIDC support for login action

### PR DESCRIPTION
**Pull request by BALAGA-GAYATRI**

### CODE REVIEW

File: src/main.ts [hunk 1]
```diff
@@ -39,31 +39,61 @@ async function main() {
...
+                    resourceManagerEndpointUrl = secrets.getSecret("$.resourceManagerEndpointUrl", false);
+                }
+                else{
+                    throw new Error("Credentials are not passed for Login action."); 
+                }           
+            }
+            //If any few of the individual credentials are missing
+            else
+                throw new Error("Few credentials are missing.ClientId,tenantId are mandatory. SubscriptionId is also mandatory if allow-no-subscriptions is not set."); 
+        }
```
Comment (vineetmimrot) [inline src/main.ts:75]: Put it into a factory pattern. Your code will be much cleaner with no if/else conditions you have used below for enableOIDC check.

File: src/main.ts [hunk 2]
```diff
@@ -39,31 +39,61 @@ async function main() {
...
+        var servicePrincipalKey = null;
+        var tenantId = core.getInput('tenant-id', { required: false });
+        var subscriptionId = core.getInput('subscription-id', { required: false });
+        var resourceManagerEndpointUrl = "https://management.azure.com/";
+        var enableOIDC= true;
+        // If any of the individual credentials (clent_id, tenat_id, subscription_id) are  missing
+        if(!servicePrincipalId || !tenantId || !(subscriptionId || allowNoSubscriptionsLogin)){
+            //If all of the individual credentials (clent_id, tenat_id, subscription_id) are missing in workflow inputs, checking for creds object.
+            if(!servicePrincipalId && !tenantId && (!subscriptionId || !allowNoSubscriptionsLogin)){
+                console.log('checking line 55 condition..')
```
Comment (vineetmimrot) [inline src/main.ts:58]: Remove all unnecessary console logs and also convert useful ones to debug logs unless required.

File: lib/main.js [hunk 3]
```diff
@@ -53,26 +66,66 @@ function main() {
...
-                throw new Error("Not all values are present in the creds object. Ensure clientId, clientSecret and tenantId are supplied.");
+            //Check for the credentials in individual parameters in the workflow.
+            var servicePrincipalId = core.getInput('client-id', { required: false });
+            ;
+            var servicePrincipalKey = null;
+            var tenantId = core.getInput('tenant-id', { required: false });
+            var subscriptionId = core.getInput('subscription-id', { required: false });
+            var resourceManagerEndpointUrl = "https://management.azure.com/";
+            var enableOIDC = true;
+            // If any of the individual credentials (clent_id, tenat_id, subscription_id) are  missing
```
Comment (kaverma) [inline lib/main.js:82]: nit: please add newlines for formatting, readability.

Comment (BALAGA-GAYATRI) [reply lib/main.js:83]: added

File: src/main.ts [hunk 4]
```diff
@@ -83,41 +121,64 @@ async function main() {
...
-                "--allow-no-subscriptions",
-                "--service-principal",
-                "-u", servicePrincipalId,
-                "-p", servicePrincipalKey,
-                "--tenant", tenantId
-            ];
+            var args = [];
+            if (enableOIDC) {
+                args = [
+                    "--allow-no-subscriptions",
```
Comment (vineetmimrot) [inline src/main.ts:146]: You can create a base args array that contains common keys and based on the specific parameter values you can append to it.
Like if `allowNoSubscriptionsLogin` is true -> append `"--allow-no-subscriptions"` to `args`, same for other keys.

File: src/main.ts [hunk 5]
```diff
@@ -161,18 +222,19 @@ async function main() {
...
+    command: string,
+    silent?: boolean,
+    execOptions: any = {},
     args: any = []) {
-    
+
     execOptions.silent = !!silent;
     try {
-        await exec.exec(`"${azPath}" ${command}`, args,  execOptions); 
+        core.debug(args);
```
[the remark this answers is not in the document]
Comment (BALAGA-GAYATRI) [reply src/main.ts:232]: I have checked it.They are getting masked. The reason putting these here is user will get to know the cli arguments as we are having two different kinds.(will be helpful to debug )
```
 ##[debug]["--service-principal","-u","***","-p","***","--tenant","***"]
##[debug]["--subscription","***"]
```

File: src/main.ts (hunk 4 above)
Comment (BALAGA-GAYATRI) [reply src/main.ts]: Done!✅

File: lib/main.js [hunk 6]
```diff
@@ -105,23 +158,27 @@ function main() {
             yield executeAzCliCommand(`cloud set -n "${environment}"`, false);
             console.log(`Done setting cloud: "${environment}"`);
             // Attempting Az cli login
+            var commonArgs = ["--service-principal",
+                "-u", servicePrincipalId,
+                "--tenant", tenantId
+            ];
             if (allowNoSubscriptionsLogin) {
```
Comment (vineetmimrot) [inline lib/main.js:165]:
```suggestion
            if (allowNoSubscriptionsLogin) {
                 commonArgs = commonArgs.concat("--allow-no-subscriptions");
            }
            
            if (enableOIDC) {
               commonArgs = commonArgs.concat("--federated-token", idToken);
            }
            else {
                commonArgs = commonArgs.concat("-p", servicePrincipalKey);
            }
            yield executeAzCliCommand(`login`, true, {}, commonArgs);
```

**Pull request merged**